dbms/src/Storages/KVStore/Utils/AsyncTasks.h

```diff
@@ -14,110 +14,355 @@
 
 #pragma once
 
+#include <Common/FmtUtils.h>
+#include <Common/Logger.h>
+#include <Common/SyncPoint/SyncPoint.h>
 #include <Common/UniThreadPool.h>
 
 #include <future>
+#include <magic_enum.hpp>
+#include <unordered_map>
 
 namespace DB
 {
+
+namespace AsyncTaskHelper
+{
+// It's guaranteed a task is no longer accessible once canceled or has its result fetched.
+// NotScheduled -> InQueue, Running, NotScheduled
+// InQueue -> Running
+// Running -> Finished, NotScheduled(canceled)
+// Finished -> NotScheduled(fetched)
+// NOTE: magic_enum can't work properly with enums in template classes due to compiler opts.
+enum class TaskState
+{
+    NotScheduled,
+    InQueue,
+    Running,
+    Finished,
+};
+} // namespace AsyncTaskHelper
+
 template <typename Key, typename Func, typename R>
 struct AsyncTasks
 {
     // We use a big queue to cache, to reduce add task failures.
     explicit AsyncTasks(uint64_t pool_size, uint64_t free_pool_size, uint64_t queue_size)
         : thread_pool(std::make_unique<ThreadPool>(pool_size, free_pool_size, queue_size))
+        , log(DB::Logger::get())
     {}
 
-    bool discardTask(Key k)
+    ~AsyncTasks() { LOG_INFO(log, "Pending {} tasks when destructing", count()); }
+
+    using TaskState = AsyncTaskHelper::TaskState;
+
+    struct CancelHandle;
+    using CancelHandlePtr = std::shared_ptr<CancelHandle>;
+    struct CancelHandle
+    {
+        CancelHandle() = default;
+        CancelHandle(const CancelHandle &) = delete;
+
+        bool isCanceled() const { return inner.load(); }
+
+        bool blockedWaitFor(std::chrono::duration<double, std::milli> timeout)
+        {
+            // The task could be canceled before running.
+            if (isCanceled())
+                return true;
+            std::unique_lock<std::mutex> lock(mut);
+            cv.wait_for(lock, timeout, [&]() { return isCanceled(); });
+            return isCanceled();
+        }
+
+        static CancelHandlePtr genAlreadyCanceled() noexcept
+        {
+            auto h = std::make_shared<CancelHandle>();
+            h->inner.store(true);
+            return h;
+        }
+
+    private:
+        // Make this private to forbid called by multiple consumers.
+        void doCancel()
+        {
+            // Use lock here to prevent losing signal.
+            std::scoped_lock<std::mutex> lock(mut);
+            inner.store(true);
+            cv.notify_all();
+        }
```

Comment:
nit: Why not acquiring lock will lead to signal lost?

Reply:
https://www.calvinneo.com/2017/12/28/Concurrency-Programming-Compare/#:~:text=%E5%8F%AF%E8%83%BD%E5%9C%A8%20wait%20%E9%83%A8%E5%88%86%E5%9C%A8%E6%A3%80%E6%B5%8B%E5%88%B0%20condition%20%E4%B8%8D%E6%BB%A1%E8%B6%B3%E5%87%86%E5%A4%87%20pthread_cond_wait%20%E5%89%8D%EF%BC%8C%E5%8F%A6%E4%B8%80%E4%B8%AA%E7%BA%BF%E7%A8%8B%E4%BF%AE%E6%94%B9%20condition%20%E5%B9%B6%20pthread_cond_signal%EF%BC%8C%E8%BF%99%E6%A0%B7%E8%BF%9B%E5%85%A5%20pthread_cond_wait%20%E7%9A%84%20wait%20%E9%83%A8%E5%88%86%E4%BB%A3%E7%A0%81%E5%B0%B1%E4%BC%9A%E4%B8%A2%E5%A4%B1%E8%BF%99%E6%AC%A1%E7%9A%84%20signal